### PR TITLE
feat: Add Immutable Seaport and Immutable Signable Zone to repository

### PR DESCRIPTION
**Pull request by Sam-Jeston**

### CODE REVIEW

File: contracts/ImmutableSeaport.sol [hunk 1]
```diff
@@ -0,0 +1,620 @@
...
+ * @custom:version 1.5
+ * @notice Seaport is a generalized native token/ERC20/ERC721/ERC1155
+ *         marketplace with lightweight methods for common routes as well as
+ *         more flexible methods for composing advanced orders or groups of
+ *         orders. Each order contains an arbitrary number of items that may be
+ *         spent (the "offer") along with an arbitrary number of items that must
+ *         be received back by the indicated recipients (the "consideration").
+ */
+contract ImmutableSeaport is Consideration, Ownable2Step {
+// contract ImmutableSeaport is Consideration {
```
Comment (lfportal) [inline contracts/ImmutableSeaport.sol:30]:
```suggestion
```

File: contracts/ImmutableSeaport.sol [hunk 2]
```diff
@@ -0,0 +1,620 @@
...
+    // Mapping to store valid ImmutableZones - this allows for multiple Zones
+    // to be active at the same time, and can be expired or added on demand.
+    mapping(address => bool) public immutableZones;
```
Comment (lfportal) [inline contracts/ImmutableSeaport.sol:33]: **Nit Suggestion:** Whilst this is Immutable's Seaport fork, the `immutable*` prefix doesn't help describe the intended behaviour. Could we change this to something like `allowedZones`?

File: contracts/ImmutableSeaport.sol [hunk 3]
```diff
@@ -0,0 +1,620 @@
...
+
+    error OrderNotRestricted();
+    error InvalidZone(address zone);
```
Comment (lfportal) [inline contracts/ImmutableSeaport.sol:36]: **Suggestion:** More of a question - Is it a recommended pattern to contain `error` and `event` definitions in a separate `.sol` file? I've seen this pattern in many contracts but haven't thought of an obvious reason for doing so.

File: contracts/ImmutableSeaport.sol [hunk 4]
```diff
@@ -0,0 +1,620 @@
...
+     *         separator during deployment.
+     *
+     * @param conduitController A contract that deploys conduits, or proxies
+     *                          that may optionally be used to transfer approved
+     *                          ERC20/721/1155 tokens.
+     */
+    constructor(address conduitController) Consideration(conduitController) Ownable2Step() {}
+
+    // Mark a zone address as active/valid
+    function addImmutableZone(
```
Comment (lfportal) [inline contracts/ImmutableSeaport.sol:49]: **Suggestion:** Not a standard or anything, but this is functionally similar to `ERC721.setApprovalForAll`. Could we follow a similar function signature and name? Something like `setAllowedZone(address zone, bool allowed)` and drop the `removeImmutableZone` function altogether?

File: contracts/conduit/ImmutableConduitController.sol [hunk 5]
```diff
@@ -0,0 +1,7 @@
+// Copyright (c) Immutable Pty Ltd 2018 - 2023
+// SPDX-License-Identifier: Apache-2
+pragma solidity 0.8.14;
+
+import { ConduitController  } from "seaport-core/src/conduit/ConduitController.sol";
+
+contract ImmutableConduitController is ConduitController{}
```
Comment (lfportal) [inline contracts/conduit/ImmutableConduitController.sol:7]: **Question:** If we are not making modifications, could we drop this and import only to deploy `ConduitController`?

File: contracts/validators/ImmutableReadOnlyOrderValidator.sol [hunk 6]
```diff
@@ -0,0 +1,7 @@
+// Copyright (c) Immutable Pty Ltd 2018 - 2023
+// SPDX-License-Identifier: Apache-2
+pragma solidity 0.8.17;
+
+import { ReadOnlyOrderValidator } from "seaport/contracts/helpers/order-validator/lib/ReadOnlyOrderValidator.sol";
+
+contract ImmutableReadOnlyOrderValidator is ReadOnlyOrderValidator {}
```
Comment (lfportal) [inline contracts/validators/ImmutableReadOnlyOrderValidator.sol:7]: **Question:** Similar to above, if we are not making modifications, could we drop this and import only to deploy `ReadOnlyValidator`?

File: contracts/validators/ImmutableSeaportValidator.sol [hunk 7]
```diff
@@ -0,0 +1,13 @@
+// Copyright (c) Immutable Pty Ltd 2018 - 2023
+// SPDX-License-Identifier: Apache-2
+pragma solidity 0.8.17;
+
+import { SeaportValidator } from "seaport/contracts/helpers/order-validator/SeaportValidator.sol";
+
+contract ImmutableSeaportValidator is SeaportValidator {
```
Comment (lfportal) [inline contracts/validators/ImmutableSeaportValidator.sol:7]: **Question:** Similar to above, if we are not making modifications, could we drop this and import only to deploy `SeaportValidator`?

File: contracts/validators/ImmutableSeaportValidatorHelper.sol [hunk 8]
```diff
@@ -0,0 +1,7 @@
+// Copyright (c) Immutable Pty Ltd 2018 - 2023
+// SPDX-License-Identifier: Apache-2
+pragma solidity 0.8.17;
+
+import { SeaportValidatorHelper } from "seaport/contracts/helpers/order-validator/lib/SeaportValidatorHelper.sol";
+
+contract ImmutableSeaportValidatorHelper is SeaportValidatorHelper {}
```
Comment (lfportal) [inline contracts/validators/ImmutableSeaportValidatorHelper.sol:7]: **Question:** Similar to above, if we are not making modifications, could we drop this and import only to deploy `SeaportValidatorHelper`?

File: hardhat.config.ts [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        },
+      },
+    },
+  },
+  defaultNetwork: "hardhat",
+  networks: {
+    hardhat: {
+      blockGasLimit: 30_000_000, // matches eth
+      // TODO: Have to get this removed - temporary to unblock
+      // allowUnlimitedContractSize: true,
```
Comment (lfportal) [inline hardhat.config.ts:74]: **Suggestion:** Can we action this `TODO` now?

File: package.json [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+  "version": "1.5.0",
+  "main": "contracts/Seaport.sol",
+  "repository": "https://github.com/immutable/immutable-seaport.git",
+  "author": "Immutable",
+  "license": "Apache-2.0",
+  "private": true,
+  "dependencies": {
+    "@openzeppelin/contracts": "^4.9.2",
+    "@rari-capital/solmate": "^6.4.0",
+    "seaport": "https://github.com/immutable/seaport.git#chore/compile-test-helpers"
```
Comment (lfportal) [inline package.json:12]: **Question:** Is this still necessary?

**Review by lfportal — CHANGES_REQUESTED** [review]

File: package.json (hunk 10 above)
Comment (Sam-Jeston) [reply package.json:12]: ooo good spot, ty

File: contracts/ImmutableSeaport.sol (hunk 3 above)
Comment (Sam-Jeston) [reply contracts/ImmutableSeaport.sol:36]: I'm really not sure - I did a bit of reading to try answer this, checking various "Solidity best practices" guides that people have written, none of them explicitly mention this. Happy to make an ImmutableEventsAndErrors.sol file and move them, but not sure its necessary either

File: contracts/ImmutableSeaport.sol (hunk 4 above)
Comment (Sam-Jeston) [reply contracts/ImmutableSeaport.sol:49]: Yeh I think that will be nicer, this was just the OG rollups implementation as I understand it

File: contracts/conduit/ImmutableConduitController.sol (hunk 5 above)
Comment (Sam-Jeston) [reply contracts/conduit/ImmutableConduitController.sol:7]: Indeed we could, but I actually liked how explicit the `Immutable*` wrapper is - how strongly do you feel about changing this?

File: contracts/ImmutableSeaport.sol (hunk 4 above)
Comment (Sam-Jeston) [reply contracts/ImmutableSeaport.sol]: Actioned in https://github.com/immutable/immutable-seaport/pull/1/commits/e0fad2022d0c3e4e05e6870220cdfa7e465ef6bf

File: contracts/conduit/ImmutableConduitController.sol (hunk 5 above)
Comment (Sam-Jeston) [reply contracts/conduit/ImmutableConduitController.sol]: After further consideration, I agree, and have done as suggested

**Review by lfportal — APPROVED** [review]

**Pull request merged**